### PR TITLE
Umcontroller decrypt and validation

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -103,6 +104,7 @@ type Config struct {
 	Downloader            Downloader        `json:"downloader"`
 	WorkingDir            string            `json:"workingDir"`
 	ImageStoreDir         string            `json:"imageStoreDir"`
+	ComponentsStoreDir    string            `json:"componentsStoreDir"`
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply config/config.go]: Done

File: umcontroller/umcontroller.go [hunk 2]
```diff
@@ -174,23 +183,35 @@ const (
...
+
 /***********************************************************************************************************************
  * Public
  **********************************************************************************************************************/
 
 // New creates new update managers controller.
-func New(config *config.Config, storage storage, urlTranslator URLTranslator, certProvider CertificateProvider,
-	cryptcoxontext *cryptutils.CryptoContext, insecure bool) (
+func New(config *config.Config, storage storage, certProvider CertificateProvider,
+	cryptcoxontext *cryptutils.CryptoContext, decryptContext fcrypt.CryptoContext, insecure bool) (
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply umcontroller/umcontroller.go]: Done

File: umcontroller/umcontroller.go [hunk 3]
```diff
@@ -204,38 +225,16 @@ func New(config *config.Config, storage storage, urlTranslator URLTranslator, ce
...
-			{Name: evUpdateFailed, Src: []string{statePrepareUpdate}, Dst: stateStartRevert},
-			{Name: evUpdateFailed, Src: []string{stateStartUpdate}, Dst: stateStartRevert},
-			{Name: evUpdateFailed, Src: []string{stateStartApply}, Dst: stateStartRevert},
-			{Name: evUmStateUpdated, Src: []string{stateStartRevert}, Dst: stateUpdateUmStatusOnRevert},
-			{Name: evContinue, Src: []string{stateUpdateUmStatusOnRevert}, Dst: stateStartRevert},
-			{Name: evSystemReverted, Src: []string{stateStartRevert}, Dst: stateIdle},
-
-			{Name: evConnectionTimeout, Src: []string{stateInit}, Dst: stateFaultState},
-		},
+		getFsmEvent(),
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply umcontroller/umcontroller.go]: Done

File: umcontroller/umcontroller.go [hunk 4]
```diff
@@ -119,6 +128,11 @@ type CertificateProvider interface {
 	GetCertificate(certType string, issuer []byte, serial string) (certURL, keyURL string, err error)
 }
 
+// Decrypter decrypt and validate image provider interface.
```
Comment (al1img) [inline umcontroller/umcontroller.go:131]: unclear comment: interface to decrypt and validate image?

File: umcontroller/umcontroller.go [hunk 5]
```diff
@@ -174,23 +188,35 @@ const (
...
+
 /***********************************************************************************************************************
  * Public
  **********************************************************************************************************************/
 
 // New creates new update managers controller.
-func New(config *config.Config, storage storage, urlTranslator URLTranslator, certProvider CertificateProvider,
-	cryptcoxontext *cryptutils.CryptoContext, insecure bool) (
+func New(config *config.Config, storage storage, certProvider CertificateProvider,
+	serverContex *cryptutils.CryptoContext, decryptContext Decrypter, insecure bool) (
```
Comment (al1img) [inline umcontroller/umcontroller.go:199]: Sorry, I made confusion with my multi-proposal comment. In this implementation lets name these params as: `cryptocontext` and `decrypter` accordingly.

File: umcontroller/umcontroller.go [hunk 6]
```diff
@@ -18,36 +18,43 @@
...
-	urlTranslator URLTranslator
-	eventChannel  chan umCtrlInternalMsg
-	stopChannel   chan bool
+	storage      storage
+	server       *umCtrlServer
+	eventChannel chan umCtrlInternalMsg
+	stopChannel  chan bool
+	componentDir string
+
+	decryptContext Decrypter
```
Comment (al1img) [inline umcontroller/umcontroller.go:55]: name it `decrypter`. It would be clear enough.

File: fcrypt/fcrypt.go [hunk 7]
```diff
@@ -130,6 +130,14 @@ type certificateChainInfo struct {
 	fingerprints []string
 }
 
+// CryptoContext contains necessary parameters for decryption.
```
Comment (al1img) [inline fcrypt/fcrypt.go:133]: // DecryptParams

File: imagemanager/imagemanager.go [hunk 8]
```diff
@@ -87,7 +86,7 @@ type Imagemanager struct {
 	servicesDir            string
 	tmpDir                 string
 	storage                Storage
-	cryptoContext          CryptoContext
+	decryptContext         Decrypter
```
Comment (al1img) [inline imagemanager/imagemanager.go:89]: Name it `decrypter`.

File: fcrypt/fcrypt.go [hunk 9]
```diff
@@ -220,6 +228,19 @@ func (handler *CryptoHandler) GetTLSConfig() (cfg *tls.Config, err error) {
 	return cfg, nil
 }
 
+// DecryptAndValidate decrypts and validates encrypted image.
+func (handler *CryptoHandler) DecryptAndValidate(encryptedFile, decryptedFile string, params DecryptParams) error {
+	if err := handler.decrypt(encryptedFile, decryptedFile, &params); err != nil {
+		return err
+	}
+
+	if err := handler.validateSigns(decryptedFile, &params); err != nil {
```
Comment (Eabramov84) [inline fcrypt/fcrypt.go:237]: in case of validate error should we clena p decrypted flie here?

File: umcontroller/umcontroller.go (hunk 4 above)
Comment (MykolaSolyanko) [reply umcontroller/umcontroller.go:131]: Done

File: umcontroller/umcontroller.go (hunk 5 above)
Comment (MykolaSolyanko) [reply umcontroller/umcontroller.go:199]: Done

File: umcontroller/umcontroller.go (hunk 6 above)
Comment (MykolaSolyanko) [reply umcontroller/umcontroller.go:55]: Done

File: fcrypt/fcrypt.go (hunk 7 above)
Comment (MykolaSolyanko) [reply fcrypt/fcrypt.go:133]: Done

File: imagemanager/imagemanager.go (hunk 8 above)
Comment (MykolaSolyanko) [reply imagemanager/imagemanager.go:89]: Done

File: fcrypt/fcrypt.go (hunk 9 above)
Comment (MykolaSolyanko) [reply fcrypt/fcrypt.go:237]: Done

File: umcontroller/umcontroller.go [hunk 10]
```diff
@@ -629,9 +658,27 @@ func (umCtrl *Controller) addComponentForUpdateToUm(componentInfo SystemComponen
...
 
+	entries, err := os.ReadDir(umCtrl.componentDir)
+	if err != nil {
+		log.Errorf("Can't read component directory: %v", err)
+
+		return
+	}
+
+	for _, entry := range entries {
+		componentFile := filepath.Join(umCtrl.componentDir, entry.Name())
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply umcontroller/umcontroller.go:676]: Done

File: fcrypt/fcrypt.go (hunk 9 above)
Comment (al1img) [reply fcrypt/fcrypt.go:238]: Why not remove the file in case of error? It is obvious.

Comment (MykolaSolyanko) [reply fcrypt/fcrypt.go]: Done

**Review by al1img — APPROVED** [review]

**Pull request merged**